// 0x.js/lib/test/utils/rpc.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = y[op[0] & 2 ? "return" : op[0] ? "throw" : "next"]) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [0, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var ethUtil = require("ethereumjs-util");
var request = require("request-promise-native");
var constants_1 = require("./constants");
var RPC = /** @class */ (function () {
    function RPC() {
        this.host = constants_1.constants.RPC_HOST;
        this.port = constants_1.constants.RPC_PORT;
        this.id = 0;
    }
    RPC.prototype.takeSnapshotAsync = function () {
        return __awaiter(this, void 0, void 0, function () {
            var method, params, payload, snapshotIdHex, snapshotId;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        method = 'evm_snapshot';
                        params = [];
                        payload = this.toPayload(method, params);
                        return [4 /*yield*/, this.sendAsync(payload)];
                    case 1:
                        snapshotIdHex = _a.sent();
                        snapshotId = ethUtil.bufferToInt(ethUtil.toBuffer(snapshotIdHex));
                        return [2 /*return*/, snapshotId];
                }
            });
        });
    };
    RPC.prototype.revertSnapshotAsync = function (snapshotId) {
        return __awaiter(this, void 0, void 0, function () {
            var method, params, payload, didRevert;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        method = 'evm_revert';
                        params = [snapshotId];
                        payload = this.toPayload(method, params);
                        return [4 /*yield*/, this.sendAsync(payload)];
                    case 1:
                        didRevert = _a.sent();
                        return [2 /*return*/, didRevert];
                }
            });
        });
    };
    RPC.prototype.mineBlockAsync = function () {
        return __awaiter(this, void 0, void 0, function () {
            var method, params, payload;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        method = 'evm_mine';
                        params = [];
                        payload = this.toPayload(method, params);
                        return [4 /*yield*/, this.sendAsync(payload)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    RPC.prototype.toPayload = function (method, params) {
        if (params === void 0) { params = []; }
        var payload = JSON.stringify({
            id: this.id,
            method: method,
            params: params,
        });
        this.id += 1;
        return payload;
    };
    RPC.prototype.sendAsync = function (payload) {
        return __awaiter(this, void 0, void 0, function () {
            var opts, bodyString, body;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        opts = {
                            method: 'POST',
                            uri: "http://" + this.host + ":" + this.port,
                            body: payload,
                            headers: {
                                'content-type': 'application/json',
                            },
                        };
                        return [4 /*yield*/, request(opts)];
                    case 1:
                        bodyString = _a.sent();
                        body = JSON.parse(bodyString);
                        return [2 /*return*/, body.result];
                }
            });
        });
    };
    return RPC;
}());
exports.RPC = RPC;
