// react-form-material-ui/lib/inputs/Checkbox.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = Checkbox;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Checkbox = require('material-ui/Checkbox');

var _Checkbox2 = _interopRequireDefault(_Checkbox);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function Checkbox(props) {
  var value = props.value,
      error = props.error,
      onChange = props.onChange,
      wrapperClassName = props.wrapperClassName,
      errorClassName = props.errorClassName,
      rest = _objectWithoutProperties(props, ['value', 'error', 'onChange', 'wrapperClassName', 'errorClassName']);

  var checked = Boolean(value);

  return _react2.default.createElement(
    'div',
    { className: wrapperClassName },
    _react2.default.createElement(_Checkbox2.default, _extends({
      checked: checked,
      onCheck: function onCheck(e, value) {
        return onChange(Checkbox.reversedInputValue ? !value : value, e);
      }
    }, rest)),
    error && _react2.default.createElement(
      'div',
      { className: errorClassName },
      error
    )
  );
}

Checkbox.propTypes = {
  value: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.bool]),
  error: _react.PropTypes.string,
  onChange: _react.PropTypes.func,
  wrapperClassName: _react.PropTypes.string,
  errorClassName: _react.PropTypes.string
};

Checkbox.defaultProps = {
  errorClassName: 'error'
};
