// react-form-material-ui/lib/inputs/SelectField.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

exports.default = SelectField;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _SelectField = require('material-ui/SelectField');

var _SelectField2 = _interopRequireDefault(_SelectField);

var _MenuItem = require('material-ui/MenuItem');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function SelectField(props) {
  var value = props.value,
      error = props.error,
      _onChange = props.onChange,
      options = props.options,
      includeBlank = props.includeBlank,
      children = props.children,
      rest = _objectWithoutProperties(props, ['value', 'error', 'onChange', 'options', 'includeBlank', 'children']);

  var optionItems = children;

  if (!optionItems) {
    optionItems = options || [];
    if (includeBlank && value) {
      optionItems = [{ value: '', text: props[includeBlank] }].concat(_toConsumableArray(optionItems));
    }

    optionItems = optionItems.map(function (opt, i) {
      var _ref = (typeof opt === 'undefined' ? 'undefined' : _typeof(opt)) === 'object' ? opt : { value: opt, text: opt },
          value = _ref.value,
          text = _ref.text;

      return _react2.default.createElement(_MenuItem2.default, { key: i, value: value, primaryText: text.toString() });
    });
  }

  return _react2.default.createElement(
    _SelectField2.default,
    _extends({
      value: value,
      onChange: function onChange(e, i, nextVal) {
        return value !== nextVal && _onChange(nextVal, i, e);
      },
      errorText: error
    }, rest),
    optionItems
  );
}

SelectField.propTypes = {
  value: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.number]),
  error: _react.PropTypes.string,
  onChange: _react.PropTypes.func,
  options: _react.PropTypes.arrayOf(_react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.number, _react.PropTypes.shape({
    value: _react.PropTypes.oneOfType([_react.PropTypes.string, _react.PropTypes.number]),
    text: _react.PropTypes.string
  })])),
  includeBlank: _react.PropTypes.oneOf(['floatingLabelText', 'hintText']),
  children: _react.PropTypes.node
};
